com.hazelcast.jet.impl.deployment.JetClassLoaderTest.when_processorSupplierCalled_then_contextClassLoaderSet

_master_ (commit 111323fed8ce48bf807eeaad28dda838b4ff9346)

Failed on Sonar build (Oracle JDK 11): https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-sonar/1316/testReport/com.hazelcast.jet.impl.deployment/JetClassLoaderTest/when_processorSupplierCalled_then_contextClassLoaderSet/

<details><summary>Stacktrace:</summary>

```
java.util.concurrent.CompletionException: 
com.hazelcast.jet.JetException: java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1135)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.putClassLoader(JetClassLoaderTest.java:309)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.init(JetClassLoaderTest.java:293)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$c8bea824$1(ExecutionPlan.java:456)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:537)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$10beb6ff$1(ExecutionPlan.java:455)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)

	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
	at com.hazelcast.jet.Job.join(Job.java:78)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest.when_processorSupplierCalled_then_contextClassLoaderSet(JetClassLoaderTest.java:264)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.hazelcast.jet.JetException: java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1135)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.putClassLoader(JetClassLoaderTest.java:309)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.init(JetClassLoaderTest.java:293)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$c8bea824$1(ExecutionPlan.java:456)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:537)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$10beb6ff$1(ExecutionPlan.java:455)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)

	at com.hazelcast.jet.impl.JobCoordinationService.lambda$joinSubmittedJob$11(JobCoordinationService.java:451)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:72)
	at com.hazelcast.jet.impl.MasterJobContext.setFinalResult(MasterJobContext.java:525)
	at com.hazelcast.jet.impl.MasterJobContext.lambda$finalizeJob$26(MasterJobContext.java:716)
	at com.hazelcast.jet.impl.util.ExceptionUtil.lambda$withTryCatch$0(ExceptionUtil.java:175)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251)
	at com.hazelcast.jet.impl.MasterJobContext.lambda$finalizeJob$27(MasterJobContext.java:712)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$54(JobCoordinationService.java:1297)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$55(JobCoordinationService.java:1318)
	at com.hazelcast.internal.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:64)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)
```

</details>

<details><summary>Standard output:</summary>

```
03:10:22,862  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [MetricsConfigHelper] when_processorSupplierCalled_then_contextClassLoaderSet - [LOCAL] [dev] [5.2-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
03:10:22,863  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [logo] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] 
	+       +  o    o     o     o---o o----o o      o---o     o     o----o o--o--o
	+ +   + +  |    |    / \       /  |      |     /         / \    |         |   
	+ + + + +  o----o   o   o     o   o----o |    o         o   o   o----o    |   
	+ +   + +  |    |  /     \   /    |      |     \       /     \       |    |   
	+       +  o    o o       o o---o o----o o----o o---o o       o o----o    o   
03:10:22,863  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [system] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
03:10:22,863  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [system] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Hazelcast Platform 5.2-SNAPSHOT (20220821 - 111323f) starting at [127.0.0.1]:5701
03:10:22,863  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [system] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Cluster name: dev
03:10:22,863  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [system] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Integrity Checker is disabled. Fail-fast on corrupted executables will not be performed. For more information, see the documentation for Integrity Checker.
03:10:22,863  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [system] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Jet is enabled
03:10:22,866  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [MetricsConfigHelper] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
03:10:22,869  WARN |when_processorSupplierCalled_then_contextClassLoaderSet| - [CPSubsystem] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
03:10:22,872  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [JetServiceBackend] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Setting number of cooperative threads and default parallelism to 2
03:10:22,873  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [Diagnostics] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
03:10:22,873  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [LifecycleService] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is STARTING
03:10:22,873  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [ClusterService] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - e74a2ea6-f8c7-49af-bf07-924aa3c7c48e this
]

03:10:22,873  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobCoordinationService] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Jet started scanning for jobs
03:10:22,873 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobCoordinationService] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
03:10:22,873  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [LifecycleService] when_processorSupplierCalled_then_contextClassLoaderSet - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is STARTED
03:10:22,875  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [PartitionStateManager] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Initializing cluster partition table arrangement...
03:10:22,875 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobClassLoaderService] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Creating job classLoader for job 0885-20c5-2640-0001
03:10:22,875 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobClassLoaderService] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Create processor classloader map for job 0885-20c5-2640-0001
03:10:22,875  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobCoordinationService] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Starting job 0885-20c5-2640-0001 based on submit request
03:10:22,876  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [MasterJobContext] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Didn't find any snapshot to restore for job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001
03:10:22,876  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [MasterJobContext] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Start executing job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001, execution graph in DOT format:
digraph DAG {
	"v2" [localParallelism=1];
	"v" [localParallelism=1];
}
HINT: You can use graphviz or http://viz-js.com/ to visualize the printed graph.
03:10:22,876 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [MasterJobContext] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Building execution plan for job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001
03:10:22,876 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [MasterJobContext] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Built execution plans for job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001
03:10:22,876 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [InitExecutionOperation] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Initializing execution plan for job 0885-20c5-2640-0001, execution 0885-20c5-2641-0001 from [127.0.0.1]:5701
03:10:22,877 ERROR |when_processorSupplierCalled_then_contextClassLoaderSet| - [InitExecutionOperation] hz.mystifying_fermat.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] null
java.util.ConcurrentModificationException: null
	at java.util.HashMap.computeIfAbsent(HashMap.java:1135) ~[?:?]
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.putClassLoader(JetClassLoaderTest.java:309) ~[test-classes/:?]
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.init(JetClassLoaderTest.java:293) ~[test-classes/:?]
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$c8bea824$1(ExecutionPlan.java:456) ~[classes/:?]
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31) ~[classes/:?]
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:537) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$10beb6ff$1(ExecutionPlan.java:455) ~[classes/:?]
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31) ~[classes/:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736) ~[?:?]
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217) ~[classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) ~[classes/:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102) ~[classes/:?]
03:10:22,877 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [MasterJobContext] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Init of job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001 has failures: [[127.0.0.1]:5701=java.util.ConcurrentModificationException]
03:10:22,877 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobExecutionService] hz.mystifying_fermat.async.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001 calling completeExecution because execution terminated before it started
03:10:22,877 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobExecutionService] hz.mystifying_fermat.async.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Completed execution of job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001
03:10:22,878 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobClassLoaderService] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Finish JobClassLoaders phaseCount = 0, removing classloaders for jobId=0885-20c5-2640-0001
03:10:22,878 ERROR |when_processorSupplierCalled_then_contextClassLoaderSet| - [MasterJobContext] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Execution of job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001 failed
	Start time: 2022-08-22T03:10:22.875
	Duration: 00:00:00.003
	To see additional job metrics enable JobConfig.storeMetricsAfterJobCompletion
java.util.ConcurrentModificationException: null
	at java.util.HashMap.computeIfAbsent(HashMap.java:1135) ~[?:?]
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.putClassLoader(JetClassLoaderTest.java:309) ~[test-classes/:?]
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.init(JetClassLoaderTest.java:293) ~[test-classes/:?]
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$c8bea824$1(ExecutionPlan.java:456) ~[classes/:?]
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31) ~[classes/:?]
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:537) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$10beb6ff$1(ExecutionPlan.java:455) ~[classes/:?]
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31) ~[classes/:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736) ~[?:?]
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217) ~[classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) ~[classes/:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102) ~[classes/:?]
03:10:22,879 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobCoordinationService] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] job '0885-20c5-2640-0001', execution 0885-20c5-2641-0001 is completed
03:10:22,879 ERROR |when_processorSupplierCalled_then_contextClassLoaderSet| - [JoinSubmittedJobOperation] hz.mystifying_fermat.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1135)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.putClassLoader(JetClassLoaderTest.java:309)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.init(JetClassLoaderTest.java:293)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$c8bea824$1(ExecutionPlan.java:456)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:537)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$10beb6ff$1(ExecutionPlan.java:455)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)

com.hazelcast.jet.JetException: java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1135)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.putClassLoader(JetClassLoaderTest.java:309)
	at com.hazelcast.jet.impl.deployment.JetClassLoaderTest$LeakClassLoaderPS.init(JetClassLoaderTest.java:293)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$c8bea824$1(ExecutionPlan.java:456)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:537)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$10beb6ff$1(ExecutionPlan.java:455)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)

	at com.hazelcast.jet.impl.JobCoordinationService.lambda$joinSubmittedJob$11(JobCoordinationService.java:451) ~[classes/:?]
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:930) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:907) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:72) ~[classes/:?]
	at com.hazelcast.jet.impl.MasterJobContext.setFinalResult(MasterJobContext.java:525) ~[classes/:?]
	at com.hazelcast.jet.impl.MasterJobContext.lambda$finalizeJob$26(MasterJobContext.java:716) ~[classes/:?]
	at com.hazelcast.jet.impl.util.ExceptionUtil.lambda$withTryCatch$0(ExceptionUtil.java:175) ~[classes/:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
	at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251) ~[?:?]
	at com.hazelcast.jet.impl.MasterJobContext.lambda$finalizeJob$27(MasterJobContext.java:712) ~[classes/:?]
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$54(JobCoordinationService.java:1297) ~[classes/:?]
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$55(JobCoordinationService.java:1318) ~[classes/:?]
	at com.hazelcast.internal.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:64) ~[classes/:?]
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217) ~[classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76) ~[classes/:?]
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102) ~[classes/:?]
03:10:23,473 DEBUG |when_processorSupplierCalled_then_contextClassLoaderSet| - [JobRepository] hz.mystifying_fermat.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Job cleanup took 0ms
03:10:23,518  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [JetTestSupport] Time-limited test - Terminating instanceFactory in JetTestSupport.@After
03:10:23,518  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [LifecycleService] Thread-76618 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
03:10:23,518  WARN |when_processorSupplierCalled_then_contextClassLoaderSet| - [Node] Thread-76618 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Terminating forcefully...
03:10:23,518  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [Node] Thread-76618 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Shutting down connection manager...
03:10:23,518  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [Node] Thread-76618 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Shutting down node engine...
03:10:23,520  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [NodeExtension] Thread-76618 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Destroying node NodeExtension.
03:10:23,521  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [Node] Thread-76618 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Hazelcast Shutdown is completed in 3 ms.
03:10:23,521  INFO |when_processorSupplierCalled_then_contextClassLoaderSet| - [LifecycleService] Thread-76618 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after when_processorSupplierCalled_then_contextClassLoaderSet(com.hazelcast.jet.impl.deployment.JetClassLoaderTest): BuildInfo{version='5.2-SNAPSHOT', build='20220821', buildNumber=20220821, revision=111323f, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'when_processorSupplierCalled_then_contextClassLoaderSet(com.hazelcast.jet.impl.deployment.JetClassLoaderTest):'
03:10:20, accumulated pauses: 3506 ms, max pause: 593 ms, pauses over 1000 ms: 0
```

</details>

--- a/hazelcast/src/test/java/com/hazelcast/jet/impl/deployment/JetClassLoaderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/impl/deployment/JetClassLoaderTest.java
@@ -51,6 +51,7 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -270,7 +271,7 @@ public class JetClassLoaderTest extends JetTestSupport {
 
     private static class LeakClassLoaderPS implements ProcessorSupplier {
 
-        private static final Map<String, List<ClassLoader>> classLoaders = new HashMap<>();
+        private static final Map<String, List<ClassLoader>> classLoaders = new ConcurrentHashMap<>();
         private volatile boolean blocks = false;
 
         LeakClassLoaderPS withBlock() {
@@ -329,7 +330,7 @@ public class JetClassLoaderTest extends JetTestSupport {
 
     private static class LeakClassLoaderPMS implements ProcessorMetaSupplier {
 
-        private static final Map<String, List<ClassLoader>> classLoaders = new HashMap<>();
+        private static final Map<String, List<ClassLoader>> classLoaders = new ConcurrentHashMap<>();
 
         @Override
         public void init(@Nonnull Context context) throws Exception {

--- a/hazelcast/src/test/java/com/hazelcast/jet/impl/deployment/JetClassLoaderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/impl/deployment/JetClassLoaderTest.java
@@ -48,12 +48,12 @@ import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
 
+import static java.util.Collections.synchronizedList;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.Assert.assertTrue;
 
@@ -111,9 +111,9 @@ public class JetClassLoaderTest extends JetTestSupport {
         job.resume();
         job.join();
 
-        assertClassLoaders(TargetP.classLoaders);
+        TargetP.classLoaders.assertClassLoadersTypes();
 
-        assertClassLoaderForAllMethodsChecked();
+        TargetP.classLoaders.assertClassLoaderForAllMethodsChecked();
     }
 
     /**
@@ -153,95 +153,91 @@ public class JetClassLoaderTest extends JetTestSupport {
         private volatile boolean received = false;
         private volatile boolean restored = false;
 
-        private static final Map<String, List<ClassLoader>> classLoaders = new HashMap<>();
-
-        private void putClassLoader(String methodName) {
-            List<ClassLoader> cls = classLoaders.computeIfAbsent(methodName, (key) -> new ArrayList<>());
-            cls.add(Thread.currentThread().getContextClassLoader());
-        }
+        private static final ClassLoaderMap classLoaders = new ClassLoaderMap();
 
         @Override
         public boolean isCooperative() {
-            putClassLoader("isCooperative");
+            classLoaders.putClassLoader("isCooperative");
             return super.isCooperative();
         }
 
         @Override
         protected void init(@Nonnull Context context) throws Exception {
-            putClassLoader("init");
+            classLoaders.putClassLoader("init");
             super.init(context);
         }
 
         @Override
         public void process(int ordinal, @Nonnull Inbox inbox) {
-            putClassLoader("process");
+            classLoaders.putClassLoader("process");
             super.process(ordinal, inbox);
         }
 
         @Override
         public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
-            putClassLoader("tryProcessWatermark");
+            classLoaders.putClassLoader("tryProcessWatermark");
             return super.tryProcessWatermark(watermark);
         }
 
         @Override
         public boolean tryProcess() {
-            putClassLoader("tryProcess");
+            classLoaders.putClassLoader("tryProcess");
             return super.tryProcess();
         }
 
         @Override
         public boolean completeEdge(int ordinal) {
-            putClassLoader("completeEdge");
+            classLoaders.putClassLoader("completeEdge");
             return super.completeEdge(ordinal);
         }
 
         @Override
         public boolean complete() {
-            putClassLoader("complete");
+            classLoaders.putClassLoader("complete");
             return restored && received;
         }
 
         @Override
+        @SuppressWarnings("ResultOfMethodCallIgnored")
         public boolean saveToSnapshot() {
             getOutbox().offerToSnapshot(1, 1);
-            putClassLoader("saveToSnapshot");
+            classLoaders.putClassLoader("saveToSnapshot");
             return super.saveToSnapshot();
         }
 
         @Override
         public boolean snapshotCommitPrepare() {
-            putClassLoader("snapshotCommitPrepare");
+            classLoaders.putClassLoader("snapshotCommitPrepare");
             return super.snapshotCommitPrepare();
         }
 
         @Override
         public boolean snapshotCommitFinish(boolean success) {
-            putClassLoader("snapshotCommitFinish");
+            classLoaders.putClassLoader("snapshotCommitFinish");
             return super.snapshotCommitFinish(success);
         }
 
         @Override
         protected void restoreFromSnapshot(@Nonnull Object key, @Nonnull Object value) {
-            putClassLoader("restoreFromSnapshot");
+            classLoaders.putClassLoader("restoreFromSnapshot");
         }
 
         @Override
         public boolean finishSnapshotRestore() {
             restored = true;
-            putClassLoader("finishSnapshotRestore");
+            classLoaders.putClassLoader("finishSnapshotRestore");
             return super.finishSnapshotRestore();
         }
 
         @Override
         public void close() throws Exception {
-            putClassLoader("close");
+            classLoaders.putClassLoader("close");
             super.close();
         }
 
         @Override
         public boolean closeIsCooperative() {
-            putClassLoader("closeIsCooperative");
+            classLoaders.putClassLoader("closeIsCooperative");
             return super.closeIsCooperative();
         }
 
@@ -264,14 +260,13 @@ public class JetClassLoaderTest extends JetTestSupport {
         // When
         instance.getJet().newJob(dag).join();
 
-        assertClassLoaders(LeakClassLoaderPS.classLoaders);
-        assertThat(LeakClassLoaderPS.classLoaders)
-                .containsKeys("init", "get", "close");
+        LeakClassLoaderPS.classLoaders.assertClassLoadersTypes();
+        LeakClassLoaderPS.classLoaders.assertContainsMethods("init", "get", "close");
     }
 
     private static class LeakClassLoaderPS implements ProcessorSupplier {
 
-        private static final Map<String, List<ClassLoader>> classLoaders = new ConcurrentHashMap<>();
+        private static final ClassLoaderMap classLoaders = new ClassLoaderMap();
         private volatile boolean blocks = false;
 
         LeakClassLoaderPS withBlock() {
@@ -291,24 +286,19 @@ public class JetClassLoaderTest extends JetTestSupport {
 
         @Override
         public void init(@Nonnull Context context) throws Exception {
-            putClassLoader("init");
+            classLoaders.putClassLoader("init");
         }
 
         @Nonnull
         @Override
         public Collection<? extends Processor> get(int count) {
-            putClassLoader("get");
+            classLoaders.putClassLoader("get");
             return Collections.singleton(new NoopP());
         }
 
         @Override
         public void close(@Nullable Throwable error) throws Exception {
-            putClassLoader("close");
-        }
-
-        private void putClassLoader(String methodName) {
-            List<ClassLoader> cls = classLoaders.computeIfAbsent(methodName, (key) -> new ArrayList<>());
-            cls.add(Thread.currentThread().getContextClassLoader());
+            classLoaders.putClassLoader("close");
         }
     }
 
@@ -323,61 +313,69 @@ public class JetClassLoaderTest extends JetTestSupport {
         // When
         instance.getJet().newJob(dag).join();
 
-        assertClassLoaders(LeakClassLoaderPMS.classLoaders);
-        assertThat(LeakClassLoaderPMS.classLoaders)
-                .containsKeys("init", "get", "close");
+        LeakClassLoaderPMS.classLoaders.assertClassLoadersTypes();
+        LeakClassLoaderPMS.classLoaders.assertContainsMethods("init", "get", "close");
     }
 
     private static class LeakClassLoaderPMS implements ProcessorMetaSupplier {
 
-        private static final Map<String, List<ClassLoader>> classLoaders = new ConcurrentHashMap<>();
+        private static final ClassLoaderMap classLoaders = new ClassLoaderMap();
 
         @Override
         public void init(@Nonnull Context context) throws Exception {
-            putClassLoader("init");
+            classLoaders.putClassLoader("init");
         }
 
         @Nonnull
         @Override
         public Function<? super Address, ? extends ProcessorSupplier> get(@Nonnull List<Address> addresses) {
-            putClassLoader("get");
+            classLoaders.putClassLoader("get");
             return address -> ProcessorSupplier.of(NoopP::new);
         }
 
         @Override
         public void close(@Nullable Throwable error) throws Exception {
-            putClassLoader("close");
+            classLoaders.putClassLoader("close");
         }
+    }
 
-        private void putClassLoader(String methodName) {
-            List<ClassLoader> cls = classLoaders.computeIfAbsent(methodName, (key) -> new ArrayList<>());
+    private static final class ClassLoaderMap {
+        private final Map<String, List<ClassLoader>> classLoaderMap = new ConcurrentHashMap<>();
+
+        void putClassLoader(String methodName) {
+            List<ClassLoader> cls = classLoaderMap.computeIfAbsent(methodName, (key) -> synchronizedList(new ArrayList<>()));
             cls.add(Thread.currentThread().getContextClassLoader());
         }
-    }
 
-    private void assertClassLoaderForAllMethodsChecked() {
-        // Future-proof against Processor API additions
-        Method[] methods = Processor.class.getMethods();
-        for (Method method : methods) {
-            if (Modifier.isStatic(method.getModifiers())) {
-                continue;
+        void assertClassLoadersTypes() {
+            for (Map.Entry<String, List<ClassLoader>> entry : classLoaderMap.entrySet()) {
+                List<ClassLoader> cls = entry.getValue();
+                for (ClassLoader cl : cls) {
+                    assertThat(cl)
+                            .describedAs("expecting JetClassLoader for method " + entry.getKey())
+                            .isInstanceOf(JetClassLoader.class);
+                }
             }
+        }
 
-            String name = method.getName();
-            assertThat(TargetP.classLoaders)
-                    .describedAs("method " + name + " not called")
-                    .containsKey(name);
+        void assertClassLoaderForAllMethodsChecked() {
+            // Future-proof against Processor API additions
+            Method[] methods = Processor.class.getMethods();
+            for (Method method : methods) {
+                if (Modifier.isStatic(method.getModifiers())) {
+                    continue;
+                }
+
+                String name = method.getName();
+                assertThat(classLoaderMap)
+                        .describedAs("method " + name + " not called")
+                        .containsKey(name);
+            }
+        }
+
+        void assertContainsMethods(String... methodNames) {
+            assertThat(classLoaderMap).containsKeys(methodNames);
         }
     }
 
-    private void assertClassLoaders(Map<String, List<ClassLoader>> classLoaders) {
-        for (Map.Entry<String, List<ClassLoader>> entry : classLoaders.entrySet()) {
-            List<ClassLoader> cls = entry.getValue();
-            for (ClassLoader cl : cls) {
-                assertThat(cl)
-                        .describedAs("expecting JetClassLoader for method " + entry.getKey())
-                        .isInstanceOf(JetClassLoader.class);
-            }
-        }
-    }
 }

